desc/integrals/bounce_integral.py

```diff
@@ -1,26 +1,818 @@
+    efficiently, and the fixed nature of quadrature performs better on GPUs.
+
+    Fast transforms are used where possible. Fast multipoint methods are not
+    implemented. For non-uniform interpolation, MMTs are used. It should be
+    worthwhile to use the inverse non-uniform fast transforms, so long as the
+    quadrature packs nodes at reasonable density. Fast multipoint methods are
+    preferable because they are exact, but that requires more development work.
+
+    See Also
+    --------
+    Bounce1D
+        Uses one-dimensional local spline methods for the same task.
+        Does not assume G is single-valued in θ and ζ.
+
+        Below are some advantages of ``Bounce2D`` over ``Bounce1D``.
+        The coordinates on which the root-finding must be done to map from DESC
+        to Clebsch coords is fixed to ``L*M*N``, independent of the number of
+        toroidal transits; generating the same data with DESC for input to
+        ``Bounce1D`` requires ``L*M*N*num_transit``. Furthermore, pseudo-spectral
+        interpolation of smooth functions such as |B| on each flux surface is more
+        efficient. This reduces the number of bounce integrals to be done by an
+        order of magnitude. Furthermore, C1 cubic spline interpolation is
+        inefficient to reconstruct smooth local maxima of |B|, which might be
+        important for (strongly) singular bounce integrals whose estimation
+        depends on ∂|B|/∂ζ there.
+
+    Attributes
+    ----------
+    required_names : list
+        Names in ``data_index`` required to compute bounce integrals.
+
+    """
+
+    required_names = ["B^zeta", "|B|", "iota"]
+    get_pitch_inv_quad = staticmethod(get_pitch_inv_quad)
+
+    def __init__(
+        self,
+        grid,
+        data,
+        theta,
+        N_B=32,
+        num_transit=32,
+        # TODO: Allow multiple starting labels for near-rational surfaces.
+        #  think can just concatenate along second to last axis of cheb
+        alpha=0.0,
+        quad=leggauss(32),
+        automorphism=(automorphism_sin, grad_automorphism_sin),
+        *,
+        Bref=1.0,
+        Lref=1.0,
+        is_reshaped=False,
+        check=False,
+        **kwargs,
+    ):
+        """Returns an object to compute bounce integrals.
+
+        Notes
+        -----
+        Performance may improve significantly if the spectral
+        resolutions ``m``, ``n``, ``M``, ``N``, and ``N_B`` are powers of two.
+        It is possible for the optimal ``M``, ``N``, and ``N_B`` to decrease
+        when ``m`` and ``n`` increase, though ``M`` is usually still large.
+
+        Parameters
+        ----------
+        grid : Grid
+            Tensor-product grid in (ρ, θ, ζ) with uniformly spaced nodes in
+            (2π × 2π) poloidal and toroidal coordinates.
+            Note that below shape notation defines
+            L = ``grid.num_rho``, m = ``grid.num_theta``, and n = ``grid.num_zeta``.
```

Comment:
would prefer to keep M and N for the theta/zeta resolution and use m,n for the fourier/chebyshev resolution